Feature/Building Filter

Jgom27 opened the pull request:
**Sample**

Building scene layer sample added: Building Explorer

**Summary**

Added a new sample, Building Explorer, which allows users to explore a building scene layer by toggling the visibility of different disciplines and categories, filtering by building level, and filtering by construction phase. 

**Tests**

Tested on windows 11 pc via unreal engine 5.0.

**ArcGIS Maps SDK Version**

ArcGIS Maps SDK v1.5.0

![image](https://github.com/Esri/arcgis-maps-sdk-unreal-engine-samples/assets/171297282/4a2e7cd9-717e-4059-b857-13e5994cf7f8)

Jade-JadeH commented:
Is this good for review now or later

Jgom27 commented:
> Is this good for review now or later
Good for review.

Jade-JadeH commented:
Tested different menu items in BuildingExplorer and work as expected. So good work on the UI part. 

Also found some potential issues

1. When you click from sample viewer level, the UI is incomplete for me, and at one point it also crashes

![image](https://github.com/Esri/arcgis-maps-sdk-unreal-engine-samples/assets/86364673/617e4421-47b6-4bb2-894f-92d3cce7a73e)

2. The samples in our sample viewer level is supposed to sort alphabetically (although I do notice that the order of geocoding and hit test is wrong too). 

![image](https://github.com/Esri/arcgis-maps-sdk-unreal-engine-samples/assets/86364673/7eeaddb6-9d4b-4350-884e-7a0f60635836)

3. When there's no API key, the menu still shows some default levels. 
![image](https://github.com/Esri/arcgis-maps-sdk-unreal-engine-samples/assets/86364673/ef5ca611-2c58-4cda-88c0-0627c60362e1)

4. The eye icon seems pretty blurry. I didn't know whether it's to check or uncheck in the beginning. 
![image](https://github.com/Esri/arcgis-maps-sdk-unreal-engine-samples/assets/86364673/253d10f4-c0ad-42d8-8abf-5d289efd0c94)

5. I believe here's the new D&C order. We changed it in the UI recently. The architectural one is the most used one so it makes sense to put it in the beginning. 
![image](https://github.com/Esri/arcgis-maps-sdk-unreal-engine-samples/assets/86364673/0e50b116-2a46-432e-a87c-682e536789ed)

I also have a question which is what is the use of the dustbin icon?

Jade-JadeH commented on sample_project/sample_project.uproject:
```diff
@@ -9,14 +9,19 @@
+			"Enabled": true
 		}
 	]
 }
```
is this change necessary?

Jgom27 replied:
No, accidentally pushed this.... Will revert

avr757 commented:
Nice job overall! Just some quick comments on UI components.

1. I think it would be better to leave a little bit of more margin (top and right) for the menu. Also, the logo size should line up with the title "Building Explorer".  The logo + title should aligned horizontally centered with the menu background.
![image](https://github.com/Esri/arcgis-maps-sdk-unreal-engine-samples/assets/106696949/cb8619a8-b9c5-415d-89fb-ef1b1ae2215a)

2. There should be more spacing between the Construction Phase and the slider, and the slider should be a little wider as the design. The Construction Phase Number can be moved slightly to the right to align with the Trash Can icon as the design. And as Jade mentioned, the Visibility icons seem distorted. Can you make the Structural, Electrical, Architectural fonts smaller? They should be in the same sizes as the Level or Construction Phase. Lastly, all the fonts under the Disciplines and Categories look different. They should be in "Chakra Petch" type face.
![image](https://github.com/Esri/arcgis-maps-sdk-unreal-engine-samples/assets/106696949/7ea61bbe-9041-45d1-bedd-bc64917998b2)

3. The trash bin icon should clear the level. 
4. There should be a space between Building and Explorer. And it should be in alphabetical order in the sample view list.
![image](https://github.com/Esri/arcgis-maps-sdk-unreal-engine-samples/assets/106696949/f53a9cbc-307a-4360-835d-f1dba5954baa)

mmgaw113 reviewed (changes requested):
I already mentioned the ArcGISPawn copy and Blueprint for the c++ class. Also please go through and revert the [changes](https://dev.to/lofiandcode/git-and-github-how-to-revert-a-single-file-dha) to `sample_project.uproject`. Unless we are adding plugins like OpenXR or water plugin, we don't update the Uproject to add dependencies for our plugin.

Jade-JadeH reviewed (changes requested):
see previous

mmgaw113 commented:
I'm going to take a look at it again this morning. Make sure you update the main readme. The link you use for linking to the readme for this sample won't work because it will be merged into main. So you will need to predict the link. Take a look at the others in that file.

mmgaw113 commented:
By default, the level doesn't show. After clicking on it, it shows up.
![image](https://github.com/user-attachments/assets/04417d2c-203a-4878-b136-ecfbf919be96)

mmgaw113 commented:
cc @avr757 Should we have an `Enable All` and a `Disable All` button within the UI similar to in disciplines and categories?

Jade-JadeH commented:
Thanks for address the feedback, and also change the order of hit test and geocoding : )

For point 3, I still see the total number of construction phase. This behavior needs to be aligned with arcgis online 
![image](https://github.com/user-attachments/assets/8a53e140-241e-4e41-8d26-d32d95bfcf16)

For point 4, I see eye icon are much more visible now. But I noticed that the visible eye icon has two shapes which I think should only have one
![image](https://github.com/user-attachments/assets/9f1afa5b-d6a1-4cdc-a7a4-1f7b8cf43661)

For point 5, I see in unity main we have architectural-structural-electrical whereas in sample the order is different

Jgom27 commented:
> By default, the level doesn't show. After clicking on it, it shows up. Same with construction phase
> 
> ![image](https://private-user-images.githubusercontent.com/89485705/348322242-04417d2c-203a-4878-b136-ecfbf919be96.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEwNjc5NTgsIm5iZiI6MTcyMTA2NzY1OCwicGF0aCI6Ii84OTQ4NTcwNS8zNDgzMjIyNDItMDQ0MTdkMmMtMjAzYS00ODc4LWIxMzYtZWNmYmY5MTliZTk2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE1VDE4MjA1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYwZTA3ZmM2ZTNlMTZkYTY3OGI3MzIzN2JhZDJjNTJhYmQwNTgxYWM4ZjQxODUyMzljZTA4ZTIxMjEyYTVkZjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.BW27yakN05hHl4-gRSdwSlrP3rSyd_y2UN912hlfTCg)

- The levels not showing is intentional to align with the Javascript sample:

![image](https://github.com/user-attachments/assets/75dbc078-e437-497a-8ecb-6b940446f83d)

- The construction phases not showing is now fixed.

Jgom27 commented:
> Thanks for addressing the feedback, and also changing the order of hit test and geocoding : )
> 
> For point 3, I still see the total number of construction phase. This behavior needs to be aligned with arcgis online ![image](https://private-user-images.githubusercontent.com/86364673/348437945-8a53e140-241e-4e41-8d26-d32d95bfcf16.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEwNjc5NTgsIm5iZiI6MTcyMTA2NzY1OCwicGF0aCI6Ii84NjM2NDY3My8zNDg0Mzc5NDUtOGE1M2UxNDAtMjQxZS00ZTQxLThkMjYtZDMyZDk1YmZjZjE2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE1VDE4MjA1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdiOTg2MzgzNzA3OGZhZjE1Mjc5M2E3NDYxMzBlNWVjYmI0NjJiZWMyOTI3ZDliMjliMzQ1YmJhMzRjNjQxZDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0mvMQctPP37duz4LqlfF1QOMzkaPAH_JewCBv7htE3k)
> 
> For point 4, I see eye icon are much more visible now. But I noticed that the visible eye icon has two shapes which I think should only have one ![image](https://private-user-images.githubusercontent.com/86364673/348438725-3303de10-1095-42f6-894f-82f33902e182.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEwNjc5NTgsIm5iZiI6MTcyMTA2NzY1OCwicGF0aCI6Ii84NjM2NDY3My8zNDg0Mzg3MjUtMzMwM2RlMTAtMTA5NS00MmY2LTg5NGYtODJmMzM5MDJlMTgyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE1VDE4MjA1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVmZWZkODg0ZTM4NTk2YmYzNGFmMDY5YjY1YmQyNzFlYTE1NzU5Y2MyMmI3MmE4ZGZiYzk0ZGVhMzIzMDU3NzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.wCcJ7aSGuvNU3_E3BLGw7frA5rErNdxqYYxZckkkMaE)
> 
> For point 5, I see in unity main we have architectural-structural-electrical whereas in sample the order is different

- Point 3 should be fixed now
- I have changed the order to architectural-structural-electrical. The Javascript sample does use alphabetical order, but I think being in line with the plugin is probably a good idea. @avr757 any thoughts?
- I cannot recreate the visible icon shape discrepancy. This is very odd.

avr757 commented:
@Jgom27 
One last thing, let's make sure there is a little more space between the logo and building exporerer, and align the logo up with level below:

<img width="471" alt="image" src="https://github.com/user-attachments/assets/2fc535ec-374c-4bf7-ab15-d71865a7b186">



So something like this:

<img width="608" alt="image" src="https://github.com/user-attachments/assets/f5efd257-b89b-4ac3-aecb-33a46982ba74">

ZackAllen commented on sample_project/Source/sample_project/BuildingExplorer/SimpleBuildingSceneLayerActor.cpp:
```diff
@@ -0,0 +1,340 @@
+
+				for (const auto& Layer : AllLayers)
+				{
+					if (Layer.Type == 7)
```
```suggestion
					if (Layer.Type == EArcGISLayerType::ArcGISBuildingSceneLayer)
```

ZackAllen commented on sample_project/Source/sample_project/BuildingExplorer/SimpleBuildingSceneLayerActor.cpp:
```diff
@@ -0,0 +1,340 @@
+			LoadStatus = TEXT("Loaded");
+		}
+	});
+	if (LoadStatus == TEXT("Failed"))
```
Don't use a string comparison for this, add a boolean in the function.

ZackAllen commented on sample_project/Source/sample_project/BuildingExplorer/SimpleBuildingSceneLayerActor.cpp:
```diff
@@ -0,0 +1,340 @@
+
+	// Define the order
+	TMap<FString, int32> DisciplineOrder;
+	DisciplineOrder.Add(TEXT("Architectural"), 0);
```
Architectural
Structural
Mechanical
Electrical
Piping

ZackAllen commented on sample_project/Source/sample_project/BuildingExplorer/SimpleBuildingSceneLayerActor.cpp:
```diff
@@ -0,0 +1,340 @@
+	for (auto Filter : Filters)
+	{
+		FString Name = Filter.GetName();
+		if (Name == TEXT("Filter"))
```
```suggestion
		if (Filter.GetName() == TEXT("Filter"))
```

ZackAllen commented on sample_project/Source/sample_project/BuildingExplorer/SimpleBuildingSceneLayerActor.cpp:
```diff
@@ -0,0 +1,340 @@
+void ASimpleBuildingSceneLayerActor::PopulateSublayerMaps(FString option, bool bVisible)
+{
+	// Search for given discipline/category
+	if (BuildingSceneLayer)
```
```suggestion
	if (!BuildingSceneLayer)
	{
	    return;
	}
```

ZackAllen commented on sample_project/Source/sample_project/BuildingExplorer/SimpleBuildingSceneLayerActor.cpp:
```diff
@@ -0,0 +1,340 @@
+		const auto& FirstLayers = BuildingSceneLayer->GetSublayers();
+		for (const auto& FirstSubLayer : FirstLayers)
+		{
+			if (FirstSubLayer.GetName() == TEXT("Full Model"))
```
You are enabling all of the full model sublayers, but you never disable the Overview layer.

ZackAllen commented on sample_project/Source/sample_project/BuildingExplorer/SimpleBuildingSceneLayerActor.cpp:
```diff
@@ -0,0 +1,340 @@
+	}
+
+	// Create the where clauses
+	FString BuildingLevelClause = FString::Printf(TEXT("BldgLevel in %s"), *BuildingLevels);
```
We user `auto` for these.

ZackAllen reviewed (changes requested):
Some initial feedback, I'll look more Monday.

arash-sh commented on sample_project/Source/sample_project/BuildingExplorer/SimpleBuildingSceneLayerActor.cpp:
```diff
@@ -0,0 +1,354 @@
+/* Copyright 2023 Esri
```
I believe we are supposed to use the current year for all new files

arash-sh commented:
Nice sample! The level and construction phase options don't seem to work when the level is opened directly (without sample viewer).

arash-sh commented:
> * The levels not showing is intentional to align with the Javascript sample:

I think in that sample, only the level selected is shown (not all levels up to that number). Should the label ("Level") be renamed to make the behavior more clear? 
Also, would it make sense to show the maximum number of levels and the highest construction phase by default?

Jgom27 commented:
> > * The levels not showing is intentional to align with the Javascript sample:
> 
> I think in that sample, only the level selected is shown (not all levels up to that number). Should the label ("Level") be renamed to make the behavior more clear? Also, would it make sense to show the maximum number of levels and the highest construction phase by default?

Went ahead and changed the functionality to only show the level selected. We do show the highest construction phase, but the highest level would not include some of the categories so having a reset button that clears the levels enables a little more functionality. However now, by default, we are showing all the levels so I think this would clear up any confusion the user might have as far as what the trash button would do.

Jgom27 commented:
> Nice sample! The level and construction phase options don't seem to work when the level is opened directly (without sample viewer).

This should be fixed now.... I did not know you could set the api key in the plugin settings.... oops.... ( I was always setting it in the map component)

mmgaw113 reviewed (approved):
Approving to not block from being merged.

arash-sh commented:
When adding new bsl, we currently add a new layer. How do you all feel about replacing the old one with the new one instead?
The main disadvantage in the current implementation is that if you add a different model of the same building (e.g., a newer version of building E bsl), both layers will be there and on top of each other.

arash-sh commented on sample_project/Content/SampleViewer/Samples/BuildingFilter/README.md:
```diff
@@ -0,0 +1,44 @@
+   - Click the play button to start the simulation.
+4. **Interact with the UI:**
+   - **Service URL:**
+     - There is a space in the UI where you can add a service URL for a building scene layer. Enter the URL to load a specific building scene layer. You can also use a local filepath for the URL.
```
```suggestion
     - There is a space in the UI where you can add a service URL for a building scene layer. Enter the URL to load a specific building scene layer. You can also use a local file path for the URL.
```

arash-sh commented on sample_project/Content/SampleViewer/Samples/BuildingFilter/README.md:
```diff
@@ -0,0 +1,44 @@
+# Building Filter
+
+Allows users to toggle the visibility of different levels, construction phases, disciplines, and categories within a building scene layer. This sample demonstrates how to explore building scene layers and filter based on various criteria.
```
```suggestion
This sample demonstrates how to explore building scene layers. User can toggle the visibility of different levels, construction phases, disciplines, and categories within the building scene layer.
```

arash-sh replied:
Not really important but I feel the main thing is exploring bsls
